### PR TITLE
Adds atom linker component/mapping helper/admin browser

### PR DESCRIPTION
**Pull request by Cyberboss**
This is a component to help objects establish links at mapping time without the use of global vars. Plenty of things already do this using globals, goal is to eventually port them all. Uses a simple string id system set using `map_text`. Links can be checked in LateInitialize using `GET_COMPONENT(/datum/component/atom_linker)` and calling `GetLinks()` (includes `src`). Adds a debug verb to view all link groups and vv them

Things to port:

~~- [ ] Cycle links~~
~~- [ ] Telecomms stuff~~

>no commitment

### CODE REVIEW

**Comment by optimumtact** [comment]
radio machinery uses heavy linking at runtime

**Comment by Cyberboss** [comment]
1. Not porting cycle links since there's so many of them and they work _o k_ as is.
2. Lets' just get this merged for now. Force new things to use it (goof's PR) and slowly port the stuff as it comes along

**Comment by Cyberboss** [comment]
What I'll actually probably do is change cyclelink helpers to use atom linkers

**Comment by Cyberboss** [comment]
@optimumtact, care to point me to an example? My grepping is weak

**Comment by optimumtact** [comment]
https://github.com/tgstation/tgstation/blob/master/code/game/machinery/telecomms/telecomunications.dm#L91

**Comment by Iamgoofball** [comment]
@Cyberboss hey hey hey this needs done still

**Comment by Cyberboss** [comment]
I'm done with this for now, the component and mapping helper works and I'm too busy to handle porting anything atm.

**Comment by optimumtact** [comment]
it would help if it compiled

File: code/datums/components/atom_linker.dm [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		var/datum/component/atom_linker/al = I
+		if(al.id == id)
+			return al.linked_components.Copy()
+
+/proc/GetLinkedAtomsWithId(id)
+	var/datum/component/atom_linker/al
+	//remove compiler warning
+	pass(al)
+	var/list/linker_lists = al.linker_lists
+	return linker_lists[id]
```
Comment (SpaceManiac) [inline code/datums/components/atom_linker.dm]: Just put it on GLOB if it's supposed to be exposed.

File: code/modules/mapping/mapping_helpers.dm [hunk 2]
```diff
@@ -178,13 +178,13 @@ GLOBAL_LIST_EMPTY(z_is_planet)
 			continue
 		if(target_type && !istype(A,target_type))
 			continue
-		var/cargs = build_args()
+		var/cargs = list(component_type) + build_args()
 		A.AddComponent(arglist(cargs))
 		qdel(src)
 		return
```
Comment (SpaceManiac) [inline code/modules/mapping/mapping_helpers.dm]: src is not qdel'd if it doesn't find anything

File: code/datums/components/atom_linker.dm [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	linked_components -= src
+	linked_components = null
+	if(!linked_components.len)
+		linker_lists[id] = null
+	return ..()
+
+/datum/component/atom_linker/proc/GetLinks()
+	return linked_components.Copy()
+
+/atom/proc/GetLinkedAtoms(id)
```
Comment (SpaceManiac) [inline code/datums/components/atom_linker.dm]: When would you call this already knowing what id to pass in, if the id is intended to be defined entirely by the component?

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

**Comment by optimumtact** [comment]
@Cyberboss you shit cunt

**Pull request closed without merging**